pmd-java/src/main/java/net/sourceforge/pmd/lang/java/typeresolution/ClassTypeResolver.java

```diff
@@ -624,6 +627,83 @@ public Object visit(ASTType node, Object data) {
         return data;
     }
 
+    private void populateVariableDeclaratorFromType(ASTLocalVariableDeclaration node) {
+        ASTType type = node.getTypeNode();
+        // also assign this type to VariableDeclarator and VariableDeclaratorId
+        TypeNode var = node.getFirstChildOfType(ASTVariableDeclarator.class);
+        if (var != null) {
+            var.setTypeDefinition(type.getTypeDefinition());
+            var = var.getFirstChildOfType(ASTVariableDeclaratorId.class);
+        }
+        if (var != null) {
+            var.setTypeDefinition(type.getTypeDefinition());
+        }
+    }
+
+    @Override
+    public Object visit(ASTLocalVariableDeclaration node, Object data) {
+        super.visit(node, data);
+        // resolve "var" types: Upward projection of the type of the initializer expression
+        ASTType type = node.getFirstChildOfType(ASTType.class);
```

Comment:
You could do `node.getTypeNode()`. Having a `isTypeInferred` on ASTLocalVariableDeclaration would help when there is no Type node. The basic contract would be the same as for `VariableDeclaratorId`: `decl.isTypeInferred() implies decl.getTypeNode() == null`

Reply:
I get your point with the ASTType being null.
Currently ASTLocalVariableDeclaration is not a TypeNode, which might be correct, since you can define multiple variables. However, they all need to have the same type, so we *could* make ASTLocalVariableDeclaration to be a `TypeNode`, too.

I've seen so far the flag `isTypeInferred` more or less as a implementation/information detail, that we can lookup, if we notice somehow "we resolved the wrong type, where does it come from? inferred from somewhere else".

I can give that a try, but return `null` or return ASTType with "null" children is equivalent problematic, since you always need to check anyways...